File: Saber/app/src/main/java/com/pathfinder/saber/MainActivity.java
package com.pathfinder.saber;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraManager;
import android.os.Bundle;
import android.support.constraint.ConstraintLayout;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.text.Layout;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.CompoundButton;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.ToggleButton;

public class MainActivity extends AppCompatActivity {

    private CameraManager mCameraManager;
    private String mCameraId;
    private ToggleButton toggleButton;
    Window window;
    ConstraintLayout layout;
    TextView subhead;
    Animation fadeoff, fadeon;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        getWindow().getDecorView().setBackgroundColor(getColor(R.color.teal));
        window = this.getWindow();
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
        window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        window.setStatusBarColor(this.getResources().getColor(R.color.teal));
        window.setNavigationBarColor(this.getResources().getColor(R.color.teal));


        fadeoff = AnimationUtils.loadAnimation(getApplicationContext(),R.anim.fadelitoff);
        fadeon = AnimationUtils.loadAnimation(getApplicationContext(),R.anim.fadeliton);
        subhead = findViewById(R.id.subheading);
        ImageButton info = findViewById(R.id.infoButt);
        info.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(MainActivity.this, About.class);
                startActivity(intent);
                overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
            }
        });

        boolean isFlashAvailable = getApplicationContext().getPackageManager()
                .hasSystemFeature(PackageManager.FEATURE_CAMERA_FLASH);

        if (!isFlashAvailable) {
            showNoFlashError();
        }

        toggleButton = findViewById(R.id.toggleButton);
        mCameraManager = (CameraManager) getSystemService(Context.CAMERA_SERVICE);
        try {
            mCameraId = mCameraManager.getCameraIdList()[0];
        } catch (CameraAccessException e) {
            e.printStackTrace();
        }

        try {
            mCameraManager.setTorchMode(mCameraId, false);
        } catch (CameraAccessException e) {
            e.printStackTrace();
        }

        toggleButton.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                //toggleButton.startAnimation(fadeoff);
                switchFlashLight(isChecked);
                toggleImag(isChecked);
            }
        });
    }

    public void showNoFlashError() {
        AlertDialog alert = new AlertDialog.Builder(this)
                .create();
        alert.setTitle("Error");
        alert.setMessage("Flash hardware not found.");
        alert.setButton(DialogInterface.BUTTON_POSITIVE, "Okay", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int which) {
                finish();
            }
        });
        alert.show();
    }

    public void switchFlashLight(boolean status) {
        try {

            mCameraManager.setTorchMode(mCameraId, status);
        } catch (CameraAccessException e) {
            e.printStackTrace();
        }
    }

    public void toggleImag(boolean status){
        if(status){
            window.setStatusBarColor(this.getResources().getColor(R.color.colorPrimaryDark));
            window.setNavigationBarColor(this.getResources().getColor(R.color.colorPrimaryDark));
            getWindow().getDecorView().setBackgroundColor(getColor(R.color.colorPrimaryDark));
            subhead.startAnimation(fadeon);
            subhead.setText(R.string.textlit);
            toggleButton.startAnimation(fadeon);
            toggleButton.setButtonDrawable(R.drawable.roomlitpartsc);

        } else{
            window.setStatusBarColor(this.getResources().getColor(R.color.teal));
            window.setNavigationBarColor(this.getResources().getColor(R.color.teal));
            getWindow().getDecorView().setBackgroundColor(getColor(R.color.teal));
            subhead.startAnimation(fadeon);
            subhead.setText(R.string.textunlit);
            toggleButton.startAnimation(fadeon);
            toggleButton.setButtonDrawable(R.drawable.ic_roomunlit);
        }
    }

}

File: Saber/app/src/main/java/com/pathfinder/saber/MainActivity.kt
package com.pathfinder.saber

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.hardware.camera2.CameraAccessException
import android.hardware.camera2.CameraManager
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.content.ContextCompat
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.InterstitialAd
import com.google.android.gms.ads.MobileAds
import com.google.android.material.snackbar.Snackbar


class MainActivity : AppCompatActivity() {
    private var mCameraManager: CameraManager? = null
    private var mCameraId: String? = null
    private lateinit var flash: ImageView
    private var subhead: TextView? = null
    private var fadeoff: Animation? = null
    private var fadeon: Animation? = null
    private var state = true
    lateinit var mInterstitialAd: InterstitialAd
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
        setContentView(R.layout.activity_main)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        }
        window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            window.attributes.layoutInDisplayCutoutMode =
                    WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_SHORT_EDGES
        }
        MobileAds.initialize(this, "ca-app-pub-9252793240012402~1634633493")
        mInterstitialAd = InterstitialAd(this)

        //TODO: replace with actual adUnitId ca-app-pub-9252793240012402/9129980131
        mInterstitialAd.adUnitId = "ca-app-pub-3940256099942544/1033173712"
        mInterstitialAd.loadAd(AdRequest.Builder().build())
        fadeoff = AnimationUtils.loadAnimation(applicationContext, R.anim.fadelitoff)
        fadeon = AnimationUtils.loadAnimation(applicationContext, R.anim.fadeliton)
        subhead = findViewById(R.id.subheading)
        flash = findViewById(R.id.flashToggle)

        findViewById<ImageButton>(R.id.infoButt).setOnClickListener {
            startActivity(Intent(this@MainActivity, About::class.java))
            overridePendingTransition(R.anim.fadeliton, R.anim.fadelitoff)
        }
        var isFlashAvailable = applicationContext.packageManager
                .hasSystemFeature(PackageManager.FEATURE_CAMERA_FLASH)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            mCameraManager = getSystemService(Context.CAMERA_SERVICE) as CameraManager
            try {
                mCameraId = mCameraManager?.cameraIdList?.get(0)
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    mCameraManager?.setTorchMode(mCameraId!!, false)
                } else isFlashAvailable = false
            } catch (e: CameraAccessException) {
                isFlashAvailable = false
            }
        } else isFlashAvailable = false
        flash.setOnClickListener {
            if (!isFlashAvailable) {
                showNoFlashError()
            } else {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    switchFlashLight(state)
                } else showNoFlashError()
            }
        }
    }

    private fun adListener() {
        mInterstitialAd.adListener = object : AdListener() {
            override fun onAdClosed() {
                Snackbar.make(findViewById(R.id.mainActivity), getString(R.string.remove_ads), 8000)
                        .setAction("Show me") {
                            Toast.makeText(applicationContext, getString(R.string.unlock_premium), Toast.LENGTH_LONG).show()
                            startActivity(Intent(this@MainActivity, About::class.java))
                            overridePendingTransition(R.anim.fadeliton, R.anim.fadelitoff)
                        }
            }
        }
    }
    override fun onRestart() {
        super.onRestart()
        if (mInterstitialAd.isLoaded) {
            mInterstitialAd.show()
            adListener()
        }
    }
    private fun showNoFlashError() {
        val alert = AlertDialog.Builder(this).create()
        alert.setTitle(getString(R.string.flash_error_title))
        alert.setMessage(getString(R.string.flash_error_message))
        alert.setIcon(ContextCompat.getDrawable(this, R.drawable.ic_launcher_foreground))
        alert.setButton(DialogInterface.BUTTON_POSITIVE, getString(R.string.sad)) { _, _ -> finish() }
        alert.show()
    }

    @RequiresApi(Build.VERSION_CODES.M)
    private fun switchFlashLight(status: Boolean) {
        try {
            mCameraManager?.setTorchMode(mCameraId!!, status)
            if (status) {
                window.statusBarColor = ContextCompat.getColor(this, R.color.colorPrimaryDark)
                window.navigationBarColor = ContextCompat.getColor(this, R.color.colorPrimaryDark)
                subhead!!.startAnimation(fadeon)
                subhead!!.setText(R.string.textlit)
                flash.startAnimation(fadeon)
                flash.setBackgroundResource(R.drawable.square_btn_teal_dark)
                flash.setImageDrawable(getDrawable(R.drawable.roomlit))
            } else {
                window.statusBarColor = ContextCompat.getColor(this, R.color.teal)
                window.navigationBarColor = ContextCompat.getColor(this, R.color.teal)
                subhead!!.startAnimation(fadeon)
                subhead!!.setText(R.string.textunlit)
                flash.startAnimation(fadeon)
                flash.setBackgroundResource(R.drawable.square_btn_teal)
                flash.setImageDrawable(getDrawable(R.drawable.ic_roomunlit))
            }
            state = !status
        } catch (e: CameraAccessException) {
            showNoFlashError()
        }
    }
}